From bbd294d1e46b86653873ec88d49e9a15cbaae68f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Cartegnie?= <fcvlcdev@free.fr>
Date: Tue, 24 Sep 2024 18:43:08 +0700
Subject: [PATCH] codec: x265: handle X265_MAX_FRAME_THREADS on newer CPU

libx265 will bail out when hitting the limit

x265 [error]: frameNumThreads (--frame-threads) must be [0 .. X265_MAX_FRAME_THREADS)
---
 modules/codec/x265.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/modules/codec/x265.c b/modules/codec/x265.c
index 97ea3709a598..4c2b585594b2 100644
--- a/modules/codec/x265.c
+++ b/modules/codec/x265.c
@@ -36,6 +36,10 @@
 
 #include <x265.h>
 
+#ifndef X265_MAX_FRAME_THREADS
+# define X265_MAX_FRAME_THREADS 16
+#endif
+
 /*****************************************************************************
  * Module descriptor
  *****************************************************************************/
@@ -153,6 +157,8 @@ static int  Open (vlc_object_t *p_this)
     x265_param_default(param);
 
     param->frameNumThreads = vlc_GetCPUCount();
+    if(param->frameNumThreads > X265_MAX_FRAME_THREADS)
+        param->frameNumThreads = X265_MAX_FRAME_THREADS;
     param->bEnableWavefront = 0; // buggy in x265, use frame threading for now
     param->maxCUSize = 16; /* use smaller macroblock */
 
-- 
GitLab

From 83e2c3955a563b60f74f05cea57e3ab5f447c8fb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Cartegnie?= <fcvlcdev@free.fr>
Date: Tue, 24 Sep 2024 18:53:11 +0700
Subject: [PATCH] codec: x265: handle 4.0 encoding API change

refs #28799
---
 modules/codec/x265.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/modules/codec/x265.c b/modules/codec/x265.c
index 4c2b585594b..5c5f35e214d 100644
--- a/modules/codec/x265.c
+++ b/modules/codec/x265.c
@@ -71,6 +71,11 @@ static block_t *Encode(encoder_t *p_enc, picture_t *p_pict)
     x265_picture pic;
 
     x265_picture_init(&p_sys->param, &pic);
+#ifdef MAX_SCALABLE_LAYERS
+    /* Handle API changes for scalable layers output in x265 4.0 */
+    x265_picture *pics[MAX_SCALABLE_LAYERS] = {NULL};
+    pics[0] = &pic;
+#endif
 
     if (likely(p_pict)) {
         pic.pts = p_pict->date;
@@ -89,8 +94,13 @@ static block_t *Encode(encoder_t *p_enc, picture_t *p_pict)
 
     x265_nal *nal;
     uint32_t i_nal = 0;
+#ifdef MAX_SCALABLE_LAYERS
+    x265_encoder_encode(p_sys->h, &nal, &i_nal,
+                        likely(p_pict) ? &pic : NULL, pics);
+#else
     x265_encoder_encode(p_sys->h, &nal, &i_nal,
             likely(p_pict) ? &pic : NULL, &pic);
+#endif
 
     if (!i_nal)
         return NULL;
-- 
GitLab

